Add metric name validation

pgomulka opened the pull request:
This commit adds a minimum metric name validation which checks:
- metric name starts with `es.` prefix
- metric name is using `.` as a separator of elements
- metric name is using characters from a white list
- validate min number of elements = 3 elements ( prefix, group and the suffix name)
- validate max number of elements and max characters per element
- validate the suffix element in a metric name to be from the enumerated allow list

pgomulka committed:
es prefix

regex

allow list

pgomulka committed:
remove last el validation

pgomulka committed:
spotless

pgomulka committed:
fix tests

pgomulka committed:
mention about a separator

pgomulka committed:
fix test plugin

pgomulka committed:
3 elements validation

pgomulka committed:
refactor blob names

pgomulka committed:
last element validation

pgomulka committed:
fix names

pgomulka committed:
fix unit tests

pgomulka committed:
Merge remote-tracking branch 'upstream/main' into metric_name_validation

pgomulka commented:
this PR also aligns the already existing metric names to pass the validation. I am ping people who introduced the as I likely will break your dashboards in kibana.
https://github.com/elastic/elasticsearch/pull/103388/files#diff-3b0ab75398e06bcff863305ac1b6e725917476eb32d5c2ccbf09e1ed884405f7

and 
https://github.com/elastic/elasticsearch/pull/103388/files#diff-74956e42dd1188b5474ec53450471777baa197a6bceb325716e31ec00d31c3f8
@ywangd 

https://github.com/elastic/elasticsearch/pull/103388/files#diff-701cea9e04fe2f63a6e0d573100f0dff75e97c8eb299b00ded7662860c5922ef
@mattc58 


https://github.com/elastic/elasticsearch/pull/103388/files#diff-b6b5bf4262dc56bace4fb61258984a3519f448349d9efd687a64ac186418e1e5
but this is already being changed in a separate PR https://github.com/elastic/elasticsearch/pull/103089

pgomulka committed:
Merge remote-tracking branch 'upstream/main' into metric_name_validation

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

ywangd commented:
Thanks @pgomulka Let me ping more folks for awareness.

@idegtiarenko for [desired balance](https://github.com/elastic/elasticsearch/pull/103388/files#diff-3b0ab75398e06bcff863305ac1b6e725917476eb32d5c2ccbf09e1ed884405f7) metrics name changes
@JVerwolf for [blob cache](https://github.com/elastic/elasticsearch/pull/103388/files#diff-74956e42dd1188b5474ec53450471777baa197a6bceb325716e31ec00d31c3f8) metrics name changes

idegtiarenko commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java:
```diff
@@ -110,12 +110,12 @@ public DesiredBalanceReconciler(ClusterSettings clusterSettings, ThreadPool thre
         );
         undesiredAllocationsRatio = meterRegistry.registerDoubleGauge(
-            "es.allocator.desired_balance.allocations.undesired_ratio",
+            "es.allocator.desired_balance.allocations.undesired.ratio",
```
Ack. I will update dashboards once this is merged

idegtiarenko commented on modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:
```diff
@@ -0,0 +1,116 @@
+
+public class MetricNameValidator {
+    private final Pattern ALLOWED_CHARACTERS = Pattern.compile("[a-z][a-z0-9_]*");
+    private final Set<String> LAST_ELEMENT_ALLOW_LIST = Set.of(
```
Consider this to `ALLOWED_SUFFIXES` or `PERMITTED_SUFFIXES`

idegtiarenko commented on modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:
```diff
@@ -0,0 +1,116 @@
+    static final int MAX_ELEMENT_LENGTH = 30;
+
+    public void validate(String name) {
+        String[] elements = name.split("\\.");
```
Is there overall metric name length limit? (not one per element)

pgomulka replied:
yes, it is 255characters. It is being verified by an agent itself

ldematte replied:
I think it would be better to have it here too, so we can fail here consistently

pgomulka committed:
add test for max length validation

pgomulka committed:
fix test

pgomulka committed:
Merge remote-tracking branch 'upstream/main' into metric_name_validation

ldematte commented on modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:
```diff
@@ -0,0 +1,117 @@
+    public static final int MAX_ELEMENT_LENGTH = 30;
+    public static final int MAX_NUMBER_OF_ELEMENTS = 10;
+
+    public void validate(String name) {
```
Can you add javadoc, and a link to `modules/apm/NAMING.md`?

ldematte commented on server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java:
```diff
@@ -204,7 +204,10 @@ private HealthPeriodicLogger(
-        this.redMetrics.put("overall", LongGaugeMetric.create(this.meterRegistry, "es.health.overall.red", "Overall: Red", "{cluster}"));
+        this.redMetrics.put(
+            "overall",
+            LongGaugeMetric.create(this.meterRegistry, "es.health.overall.red.status", "Overall: Red", "{cluster}")
```
Trying to understand "status" here: why not "count" (0/1)?

pgomulka replied:
this is a workaround for a lack of enum/boolean support in OTEL. 
count would indicate that this value will be increasing.

ldematte commented on server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:
```diff
@@ -116,7 +116,7 @@ private void registerAsyncMetrics(MeterRegistry registry) {
         metrics.add(
             registry.registerLongGauge(
-                "es.node.stats.indices.translog.operations",
+                "es.node.stats.indices.translog.operations.count",
```
Should be singular (operation.count)

pgomulka replied:
hmm are you sure? it actually feels ok to me

ldematte replied:
Yes, from both our guidelines and https://opentelemetry.io/docs/specs/semconv/general/metrics/.
It could be either `operations` or `operation.count`, but since we decided for fixed suffixes, it should be the second.

ldematte commented on server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:
```diff
@@ -134,7 +134,7 @@ private void registerAsyncMetrics(MeterRegistry registry) {
         metrics.add(
             registry.registerLongGauge(
-                "es.node.stats.indices.translog.uncommitted_operations",
+                "es.node.stats.indices.translog.uncommitted_operations.count",
```
Same here (singular)

ldematte commented on server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:
```diff
@@ -152,7 +152,7 @@ private void registerAsyncMetrics(MeterRegistry registry) {
         metrics.add(
             registry.registerLongAsyncCounter(
-                "es.node.stats.indices.translog.earliest_last_modified_age",
+                "es.node.stats.indices.translog.earliest_last_modified_age.time",
```
Nit: isn't "age" redundant? earliest_last_modified.time looks better, IMO

pgomulka replied:
fair point, I suspect this was meant to replicate already existing metrics returned from that node-stats api
I will rename.

ldematte commented on server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:
```diff
@@ -224,7 +224,7 @@ private void registerAsyncMetrics(MeterRegistry registry) {
         metrics.add(
             registry.registerLongGauge(
-                "es.node.stats.indices.indexing.docs.current",
+                "es.node.stats.indices.indexing.docs.total",
```
Are we sure this is total and not count? I see the description names "Current". Maybe is it the description that needs update instead?

pgomulka replied:
good catch. I think this has to keep the current subgroup to indicate this is about 'in-flight indexing operations'

ldematte commented on server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:
```diff
@@ -251,7 +251,7 @@ private void registerAsyncMetrics(MeterRegistry registry) {
         metrics.add(
             registry.registerLongGauge(
-                "es.node.stats.indices.deletion.docs.current",
+                "es.node.stats.indices.deletion.docs.total",
```
Same

pgomulka replied:
I am not sure, the javadoc for   `getIndexFailedCount`  says `The number of failed indexing operations`
both `total` and `count` feel ok to me..
maybe it should not be a longGauge but asyncCounter

ldematte commented on server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:
```diff
@@ -314,7 +314,7 @@ private void registerAsyncMetrics(MeterRegistry registry) {
         metrics.add(
             registry.registerLongGauge(
-                "es.node.stats.indices.indexing.coordinating_operations.memory.size.current",
+                "es.node.stats.indices.indexing.coordinating_operations.memory.size.total",
```
Same

pgomulka replied:
I think this can be just .size

stu-elastic replied:
Yes, size is best here.

ldematte commented on server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:
```diff
@@ -359,7 +359,7 @@ private void registerAsyncMetrics(MeterRegistry registry) {
         metrics.add(
             registry.registerLongGauge(
-                "es.node.stats.indices.indexing.primary_operations.memory.size.current",
+                "es.node.stats.indices.indexing.primary_operations.memory.size.total",
```
Same as above, it is probably just `size`?

pgomulka replied:
agree

ldematte commented on server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:
```diff
@@ -368,7 +368,7 @@ private void registerAsyncMetrics(MeterRegistry registry) {
         metrics.add(
             registry.registerLongGauge(
-                "es.node.stats.indices.indexing.primary_operations.count.current",
+                "es.node.stats.indices.indexing.primary_operations.count.total",
```
Same: probably just `count`?

ldematte commented on server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:
```diff
@@ -386,7 +386,7 @@ private void registerAsyncMetrics(MeterRegistry registry) {
         metrics.add(
             registry.registerLongGauge(
-                "es.node.stats.indices.indexing.memory.limit.current",
+                "es.node.stats.indices.indexing.memory.limit.total",
```
Same: just `limit` (and it needs to be added to the suffixes too probably)

pgomulka replied:
how about memory_limit.total ?

ldematte commented on x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:
```diff
@@ -20,17 +20,17 @@ public class BlobCacheMetrics {
         this(
             meterRegistry.registerLongCounter(
-                "elasticsearch.blob_cache.miss_that_triggered_read",
+                "es.blob_cache.miss_that_triggered_read.count",
```
Is this `total` instead?

pgomulka replied:
count and total feel very similar to me. I would use `total` for async counter or gauge (reporting a cumulative value)
and count for `deltas` like here.

ldematte replied:
Count should for the current number of a fluctuating quantity, e.g. the number of tabs _currently_ opened in your browser is `browser.window.open.count`.
Total should be for an always increasing quantity, e.g. the number of times you have been oncall `oncall.total`

This looks like a total to me (we look at all the cache misses and we keep increasing the counter).

pgomulka replied:
as we discussed offline.
I will change `count` to be `current`. The intention is to use this for values of current snapshot/sample. That is for non-monotonic metrics (number of tabs can go up and down) Does not matter if it is additive or not. 
`total` will be always monotonic metrics (a counter). I will remove the word of `sum` as it might be confusing if the value reported should cumulative or delta.

ldematte commented on x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:
```diff
@@ -20,17 +20,17 @@ public class BlobCacheMetrics {
             ),
             meterRegistry.registerLongCounter(
-                "elasticsearch.blob_cache.count_of_evicted_used_regions",
+                "es.blob_cache.evicted_used_regions.count",
```
Same (possibly `total`?)

pgomulka replied:
count and total feel very similar to me. I would use total for async counter or gauge (reporting a cumulative value)
and count for deltas like here.

pgomulka replied:
changed to total

ldematte commented on x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java:
```diff
@@ -20,17 +20,17 @@ public class BlobCacheMetrics {
-                "elasticsearch.blob_cache.cache_miss_load_times",
+                "es.blob_cache.cache_miss_load.time",
                 "The timing data for populating entries in the blob store resulting from a cache miss.",
                 "count"
```
Wrong unit of measure?

stu-elastic commented on server/src/main/java/org/elasticsearch/monitor/metrics/NodeMetrics.java:
```diff
@@ -116,7 +116,7 @@ private void registerAsyncMetrics(MeterRegistry registry) {
-                "es.node.stats.indices.translog.operations",
+                "es.node.stats.indices.translog.operations.count",
                 "Number of transaction log operations.",
                 "operation",
```
Should this change to "count"?

pgomulka replied:
translong.operations.count feels addivite to me. perhaps it should be an async counter instead

pgomulka replied:
I have changed this to total.
unit here is misleading. we often use just `count` as counters are often unit-less.
I wonder if we should not  allow empty strings here

stu-elastic reviewed (approved):
lgtm once my and @ldematte's comments are addressed.

rjernst commented on modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java:
```diff
@@ -0,0 +1,117 @@
+    public static final int MAX_ELEMENT_LENGTH = 30;
+    public static final int MAX_NUMBER_OF_ELEMENTS = 10;
+
+    public void validate(String name) {
```
Could this be static? The validation methods seem to be stateless, no need to have an instance?

pgomulka replied:
I tend to start with a regular class instead of a static util. Static utilities are creating tight coupling and prevent testing in isolation.
However I have moved the use of `MetricNameValidator` to `AbstractInstrument` and refactored all the test metric names to pass the rules so there probably will never be need to test instruments in isolation of the metric name validation.

pgomulka committed:
rename of metrics

pgomulka committed:
make validator static

pgomulka committed:
rename duplicated metric name

pgomulka committed:
rename count->current for non-monotonic metrics, total (the same) for monotonic metrics (counters)

pgomulka commented:
note for reviewers( @rjernst  @stu-elastic ). Please take a look at `total` vs `current` vs `count`
I have changed the naming guide to be more clear about this
```
* Use `total` as a suffix for monotonic metrics (always increasing counter) (e.g. <code>es.indices.docs.deleted.<strong>total</strong></code>)
  * Note: even though async counter is reporting a total cumulative value, it is till monotonic.
* Use `current` to represent the non-monotonic metrics (like gauges, upDownCounters) (e.g. if we have `es.process.jvm.classes.loaded`, we will express the number of classes currently loaded by the JVM as <code>es.process.jvm.classes.loaded.<strong>count</strong></code>, and the total number of classes loaded since the JVM started as <code>es.process.jvm.classes.loaded.<strong>total</strong></code>
```
I felt like `count` was often misleading for metric names used in gauges. Hence a rename to `current`
but `current` also is not ideal see https://github.com/elastic/elasticsearch/pull/102371/files/01260a3599eef59426d505b47d4193d0e27b3339#diff-2835e151b1e7435ff4760a9263e0ecbc359597a9ff02005fc7fec099d38549e6
for instance:
 `es.thread_pool." + name + ".threads.current` (gauge)  
or
` es.thread_pool." + name + ".threads.active.current` (gauge)
those does not indicate that they are a `unit-ness` numbers.

**the alternative we discussed with @ldematte  (pretty much reverse the count and total)**
```
* Use `count` as a suffix for monotonic metrics (always increasing counter) (e.g. <code>es.indices.docs.deleted.<strong>total</strong></code>)
  * Note: even though async counter is reporting a total cumulative value, it is till monotonic.
* Use `total` to represent the non-monotonic metrics (like gauges, upDownCounters). The overall cumulative value of a measured resource at the sample time
```

pgomulka committed:
count->current

pgomulka committed:
Merge remote-tracking branch 'upstream/main' into metric_name_validation

stu-elastic commented:
Re: `count` vs `total`.  I know that the desire is to be able to differeniate between sample metrics like gauges and additive metrics like counters.

However, those two look very similar and we are adding cognitive load on authors instrumenting their code by making them pick between the two.  While differentiating between metric types is a useful goal, I do not think it's worth having two very similar units.

We are not enforcing per-instrument-type naming guidelines so we should expect regular confusion over which to pick.

I suggest dropping `total` and going with `count` and also removing `current`, as it is not a unit.

We can always add new units but it will be a backwards compatibility issue if we let one unit in and then try to remove it.

JVerwolf commented:
Question: Why do we need to manually add suffixes to metrics when this could be automated by the metric infrastructure itself? On this note, if we are going to enforce suffixes based on metric types, why not include the metric type as a suffix and remove this extra layer?  e.g. a counter could add `.counter`, a histogram `.histogram`, a gauge adds `.gauge`, etc.

rjernst commented:
> Why do we need to manually add suffixes to metrics

I think the reasoning is similar to why we don't automatically add the `es.` prefix. We want the metric name to be searchable in the codebase. So if you are looking at a dashboard, and you see the source data is `es.foo.bar.counter`, you should be able to search for that string. We have prior examples eg in xpack settings where the actual setting name is built up and therefore obfuscated, and it makes navigating to where a setting is defined difficult. We don't want to make that same mistake with metrics.

rjernst commented:
> I suggest dropping total and going with count and also removing current, as it is not a unit.

I think there is a benefit of having different suffixes for cumulative vs current values. These often appear side-by-side. For example, the number of docs indexed. We have the total of all time for the node, and the current number being indexed.

ldematte commented:
> I think there is a benefit of having different suffixes for cumulative vs current values. These often appear side-by-side. For example, the number of docs indexed. We have the total of all time for the node, and the current number being indexed.

++
While compiling the guidelines, that was exactly the scenario we depicted. You may want to have both (e.g. `es.something.cached.count` + `es.something.cached.total`

`current` was proposed as a more explicit/less ambiguous replacement for `count`

> Why do we need to manually add suffixes to metrics when this could be automated by the metric infrastructure itself?

Another reason: avoid name clashing and decouple the instrument type from the metric name.

That said, I'm happy to go either way: we can drop the difference and just keep `count` for every `unit` measure (and work around the previous case with more elaborate names, e.g. `es.something.cached.current.count` + `es.something.cached.total.count`, only when we need it).
If we go that way, I'd update the guidelines (probably in the same commit, since they now live in the repo).

pgomulka committed:
Merge remote-tracking branch 'upstream/main' into metric_name_validation

pgomulka committed:
rename count to total

pgomulka committed:
rename ml metrics

stu-elastic commented:
I'm fine with having a separate `count` vs `total` to differentiate the instrument so long as it's easy for metric authors to know when to use which suffix.

pgomulka committed:
rephrase the example of total vs current
